internal/training/grpo_engine/stream.c

```
@@ -0,0 +1,652 @@
+    se->embed_mmap_sz = (size_t)(max_offset - min_offset);
+
+    /* mmap this region */
+    se->embed_mmap = mmap(NULL, se->embed_mmap_sz, PROT_READ, MAP_SHARED,
+                         se->fd, min_offset);
+    if (se->embed_mmap == MAP_FAILED) {
+        fprintf(stderr, "stream: mmap failed for global tensors: %s\n", strerror(errno));
+        return -1;
+    }
+
+    /* Resolve pointers into mmap */
+    se->embed = (const float *)((uint8_t *)se->embed_mmap + (embed_t->offset - min_offset));
```

Cppcheck (2.21.0)

[portability] Casting between unsigned char * and const float * which have an incompatible binary data representation.

(invalidPointerCast)